### PR TITLE
Introduce compiler option to store apt generated sources

### PR DESCRIPTION
**Pull request by lptr**
Instead of specifying `-s` together with an absolute path pointing to the output directory, and breaking the relocatability of the `JavaCompile` task, users can achieve the same via `JavaCompile.options.annotationProcessorGeneratedSourcesDirectory`.

### CODE REVIEW

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=8443516

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java [hunk 1]
```diff
@@ -79,6 +84,13 @@
 
     private FileCollection annotationProcessorPath;
 
+    private final PropertyState<File> annotationProcessorGeneratedSourcesDirectory;
```
Comment (wolfs) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: Why don't we use a `DirectoryVar` here?

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java [hunk 2]
```diff
@@ -460,5 +472,37 @@ public FileCollection getAnnotationProcessorPath() {
...
+    /**
+     * Returns the directory to place source files generated by annotation compilers.
+     *
+     * @since 4.3
+     */
+    @Optional
+    @Nullable
+    @Incubating
+    @OutputDirectory
+    public File getAnnotationProcessorGeneratedSourcesDirectory() {
```
Comment (wolfs) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: When using a `DirectoryVar`, we could return it here. No need to get the `File` from the `DirectoryVar`/`PropertyState`.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java [hunk 3]
```diff
@@ -460,5 +472,37 @@ public FileCollection getAnnotationProcessorPath() {
...
+        return annotationProcessorGeneratedSourcesDirectory.getOrNull();
+    }
+
+    /**
+     * Sets the directory to place source files generated by annotation compilers.
+     *
+     * @since 4.3
+     */
+    @Incubating
+    public void setAnnotationProcessorGeneratedSourcesDirectory(File file) {
```
Comment (wolfs) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: Let's not add a setter, just a getter for `DirectoryVar`.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java [hunk 4]
```diff
@@ -75,10 +76,12 @@
  */
 @CacheableTask
 public class JavaCompile extends AbstractCompile {
-    private final CompileOptions compileOptions = new CompileOptions();
+    private final CompileOptions compileOptions;
     private JavaToolChain toolChain;
 
     public JavaCompile() {
+        CompileOptions compileOptions = getServices().get(ObjectFactory.class).newInstance(CompileOptions.class);
```
Comment (wolfs) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: Nit: I would just inline compileOptions.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 1 above)
Comment (wolfs) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: Would it make sense to add another nested object of type `AnnotationProcessorOptions` which would have the `path` and the `generatedSourcesDirectory` as properties? We could even move this object to `JavaCompile` I guess.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 2 above)
Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: I guess we want to add some `PathSensitivity` annotation here, right?

Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: Not used for output properties.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 1 above)
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: We discussed this via voice. FTR, `DirectoryVar` is currently only used in native tasks. Let's start using it when we restructure how compiler options are specified; the current API is chaotic enough as it is.

Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: Let's do something like this again when we have a plan to rework compiler options.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 3 above)
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java]: Won't use `DirectoryVar` for now.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java (hunk 4 above)
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: My pet peeve is not to mutate fields and use their values in the same method if I can avoid it. Unless you feel strongly, I'd leave this as it is.

File: subprojects/docs/src/docs/release/notes.md [hunk 5]
```diff
@@ -57,6 +57,10 @@ A problem was found with the configuration of task ':test'. Registering invalid
 
 You may now force Gradle to use rich or plain [build output](userguide/console.html#sec:console_build_output) by setting [`org.gradle.console`](userguide/build_environment.html#sec:gradle_configuration_properties) in your `gradle.properties`.
 
+### Specify output directory for source files generated by annotation processor 
+
+When using annotation processors, generated Java source files can be generated into a directory. When the directory is specified via the new `compileJava.options.annotationProcessorGeneratedSourcesDirectory` property, the generated sources are also stored and loaded from the build cache.
```
Comment (wolfs) [inline subprojects/docs/src/docs/release/notes.md]: Two recommendations:
"generated Java source files can be generated into a directory" -> "Java source may be generated into a directory specified by the `-s` compiler argument"
"the generated sources are also stored and loaded from the build cache" -> "the generated sources are also stored to and loaded from the build cache"

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java (hunk 4 above)
Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: Looks like I am so used to some patterns that I don't even notice that there is less immutability in your solution.

Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: How's there less?

File: subprojects/docs/src/docs/release/notes.md (hunk 5 above)
Comment (lptr) [reply subprojects/docs/src/docs/release/notes.md]: Done.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java (hunk 4 above)
Comment (wolfs) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java]: Sorry, less -> more.

**Pull request merged**